### PR TITLE
[MOB-14353] XDM Lifecycle Metrics

### PR DESCRIPTION
**Pull request by nporter-adbe**
## Description

Provides the XDM representation of Lifecycle data, along with an XDM metrics builder for Lifecycle data.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by nporter-adbe** [comment]
@emdobrin @kevinlind need your reviews

File: AEPLifecycle/Sources/XDM/XDMDeviceType.swift [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+import AEPServices
+import Foundation
+
+/// Represents an XDM device type, currently only supported values are mobile and tablet.
+enum XDMDeviceType: String, Encodable {
+    case mobile = "mobile"
+    case tablet = "tablet"
+    case desktop = "desktop"
+    case tvScreens = "tv screens"
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMDeviceType.swift:21]: let's keep only the values we use for now

File: AEPLifecycle/Sources/XDM/XDMDeviceType.swift [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    /// Creates a new `DeviceType` from the already known `AEPServices.DeviceType`
+    /// - Parameter servicesDeviceType: a `AEPServices.DeviceType` value
+    /// - Returns: The `servicesDeviceType` mapped to an `XDMDeviceType`, defaults to `.mobile` if no mappings found
+    static func from(servicesDeviceType: AEPServices.DeviceType) -> XDMDeviceType? {
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMDeviceType.swift:26]: the return is not optional if it always has a default

File: AEPLifecycle/Sources/XDM/XDMEnvironmentType.swift [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+import Foundation
+
+/// Represents an XDM environment type
+enum XDMEnvironmentType: String, Encodable {
+    case browser
+    case application
+    case iot
+    case external
+    case widget
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMEnvironmentType.swift:21]: Same here, let's keep only the supported values for now.

File: AEPLifecycle/Sources/XDM/XDMEnvironmentType.swift [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    /// Creates an `XDMEnvironmentType` from a run mode `String`
+    /// - Parameter runMode: The current run mode for the system as described by the `SystemInfoService`
+    /// - Returns: The matching `XDMEnvironmentType` for the run mode, nil if no matches found
+    static func from(runMode: String?) -> XDMEnvironmentType? {
+        if runMode == "Application" {
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMEnvironmentType.swift:27]: we can compare with ignorecase

File: AEPLifecycle/Sources/XDM/XDMFormatters.swift [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    /// For example, 2017-09-26T15:52:25-07:00
+    /// - Parameters:
+    ///   - Date: A timestamp and it must not be null
+    /// - Returns: The timestamp formatted to a string in the format of 'yyyy-MM-dd'T'HH:mm:ssXXX',
+    ///            or an empty string if Date  is null
+    public static func dateToISO8601String(from: Date?) -> String? {
+        if let unwrapped = from {
+            return unwrapped.asISO8601String()
+        } else {
+            return nil
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMFormatters.swift:29]: we can use guard to make it a two liner

File: AEPLifecycle/Sources/XDM/XDMFormatters.swift [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+// Unless required by applicable law or agreed to in writing, software distributed under
+// the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+// OF ANY KIND, either express or implied. See the License for the specific language
+// governing permissions and limitations under the License.
+//
+
+import Foundation
+
+/// Helper functions for XDM formatting
+public enum XDMFormatters {
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMFormatters.swift:16]: should't be public.

File: AEPLifecycle/Sources/XDM/XDMFormatters.swift [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    /// Serialize the given Date to a string formatted to an ISO 8601 date as defined in
+    /// <a href="https://tools.ietf.org/html/rfc3339#section-5.6">RFC 3339, section 5.6</a>
+    /// For example, 2017-09-26
+    /// - Parameters:
+    ///   - Date:  A timestamp and it must not be null
+    /// - Returns: The timestamp formatted to a string in the format of 'yyyy-MM-dd',
+    ///            or an empty string if Date  is null
+    public static func dateToFullDateString(from: Date?) -> String? {
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMFormatters.swift:40]: let's rename:
dateToISO8601String -> dateTimeISO8601String 
dateToFullDateString -> dateOnlyISO8601String

File: AEPLifecycle/Sources/XDM/XDMLifecycleApplication.swift [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    var closeType: XDMCloseType?
+    var id: String?
+    var installDate: Date?
+    var isClose: Bool?
+    var isInstall: Bool?
+    var isLaunch: Bool?
+    var isUpgrade: Bool?
+    var name: String?
+    var sessionLength: Int64?
+    var version: String?
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMLifecycleApplication.swift:29]: We should add docs for all these fields as er schema definition - this comment applies for all structs

File: AEPLifecycle/Sources/XDM/XDMLifecycleEnvironment.swift [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import AEPServices
+import Foundation
+
+/// Represents the Lifecycle Environment schema
+struct XDMLifecycleEnvironment: Encodable {
+    init() {}
+
+    var carrier: String?
+    var language: XDMLifecycleLanguage?
+    var operatingSystemVendor: String?
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMLifecycleEnvironment.swift:22]: not used, let's remove

File: AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        appLaunchXDMData.eventType = LifecycleConstants.XDM.EVENT_TYPE_APP_LAUNCH
+        appLaunchXDMData.timestamp = startDate
+
+        return appLaunchXDMData.asDictionary()
+    }
+
+    /// Builds the data required for the XDM Application Close event, including `XDMLifecycleApplication`
+    /// - Returns: App close event data in dictionary format
+    func buildXDMAppCloseEventData() -> [String: Any]? {
+        var appLaunchXDMData = XDMMobileLifecycleDetails()
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift:52]: should be appCloseXDMData

File: AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+    /// Adds general application information as well as details related to the type of close event
+    /// - Parameters:
+    ///   - previousAppId: application version from the previous session when the close happened
+    ///   - previousSessionInfo: previous session information to be attached to the close event
+    /// - Returns: this `XDMLifecycleMetricsBuilder` instance
+    func addAppCloseData(previousAppId: String?, previousSessionInfo: LifecycleSessionInfo) -> XDMLifecycleMetricsBuilder {
+        xdmApplicationInfoClose = XDMLifecycleApplication()
+
+        xdmApplicationInfoClose?.version = previousAppId
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift:95]: Same as Android, this should be the previousAppVersion - the appId contains app name + app versions, but we only need the version here. Let's create a task for this and reference it here

File: AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        xdmApplicationInfoClose?.isClose = true
+        xdmApplicationInfoClose?.closeType = previousSessionInfo.isCrash ? .unknown : .close
+        if let sessionLength = previousSessionInfo.sessionLength {
+            xdmApplicationInfoClose?.sessionLength = Int64(sessionLength)
+        }
+        xdmApplicationInfoClose?.name = systemInfoService.getApplicationName()
+        xdmApplicationInfoClose?.id = systemInfoService.getApplicationBundleId()
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift:102]: This info seems redundant here as we transmit this info in the app launch event

File: AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+
+        return self
+    }
+
+    /// Adds information related to the running environment, see `XDMLifecycleEnvironment`
+    /// - Returns: this `XDMLifecycleMetricsBuilder` instance
+    func addEnvironmentData() -> XDMLifecycleMetricsBuilder {
+        xdmEnvironmentInfo = XDMLifecycleEnvironment()
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift:110]: for all add* functions it looks like we reinstantiate. Is this your intention? Should we document this behavior or even reject subsequent calls to the same function in this builder?

File: AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+        xdmEnvironmentInfo?.carrier = systemInfoService.getMobileCarrierName()
+        xdmEnvironmentInfo?.type = XDMEnvironmentType.from(runMode: systemInfoService.getRunMode())
+        xdmEnvironmentInfo?.operatingSystem = systemInfoService.getOperatingSystemName()
+        xdmEnvironmentInfo?.operatingSystemVersion = systemInfoService.getOperatingSystemVersion()
+        xdmEnvironmentInfo?.language = XDMLifecycleLanguage(language: systemInfoService.getFormattedLocale())
+
+        return self
+    }
+
+    /// Adds information related to the running environment, see `XDMLifecycleEnvironment`
```
Comment (emdobrin) [inline AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift:121]: Adds information related to current device, see `XDMLifecycleDevice`

File: AEPLifecycle/Tests/XDMLifecycleDeviceTests.swift [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+import AEPServicesMocks
+import XCTest
+
+class XDMLifecycleDeviceTests: XCTestCase {
+
+    private func buildAndSetMockInfoService() {
+        let mockSystemInfoService = MockSystemInfoService()
+        mockSystemInfoService.deviceName = "test-device-name"
+        mockSystemInfoService.displayInformation = (100, 200)
+        mockSystemInfoService.orientation = .PORTRAIT
```
Comment (emdobrin) [inline AEPLifecycle/Tests/XDMLifecycleDeviceTests.swift:23]: not used

File: AEPLifecycle/Tests/XDMMobileLifecycleDetailsTests.swift [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+            "model" : "test-device-name",
+            "screenHeight" : 200,
+            "screenWidth" : 100,
+            "type" : "mobile"
+          },
+          "application" : {
+            "sessionLength" : 424394,
+            "isInstall" : true,
+            "isLaunch" : true,
+            "isUpgrade" : true,
```
Comment (emdobrin) [inline AEPLifecycle/Tests/XDMMobileLifecycleDetailsTests.swift:73]: nit: not quite possible to install and upgrade the same time 😃

File: AEPLifecycle/Sources/XDM/XDMFormatters.swift (hunk 5 above)
Comment (nporter-adbe) [reply AEPLifecycle/Sources/XDM/XDMFormatters.swift:29]: This file is copied from Edge, I'm going to leave it as-is.

File: AEPLifecycle/Sources/XDM/XDMFormatters.swift (hunk 7 above)
Comment (nporter-adbe) [reply AEPLifecycle/Sources/XDM/XDMFormatters.swift:40]: This file is copied from Edge, I'm going to leave it as-is.

File: AEPLifecycle/Tests/XDMMobileLifecycleDetailsTests.swift (hunk 16 above)
Comment (nporter-adbe) [reply AEPLifecycle/Tests/XDMMobileLifecycleDetailsTests.swift:73]: True, but this test is ensuring encoding is correct not any other functionality in the class. Going to leave this.

File: AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift (hunk 13 above)
Comment (nporter-adbe) [reply AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift]: Yeah, this is intentional, I think it makes sense as the object it creates in the function is only ever modified from within that function.

File: AEPLifecycle/Sources/LifecycleSession.swift [hunk 17]
```diff
@@ -145,4 +145,12 @@ struct LifecycleSessionInfo {
 
     /// Flag indicating whether this session crashed or not
     let isCrash: Bool
+
+    /// Computes the session length in seconds, nil if session length is invalid due to `startDate` and `pauseDate`
```
[the remark this answers is not in the document]
Comment (nporter-adbe) [reply AEPLifecycle/Sources/LifecycleSession.swift:149]: I think the doc is reasonable considering it's a computed variable.

File: AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift (hunk 12 above)
Comment (nporter-adbe) [reply AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift:102]: Aren't we doing that here? https://git.corp.adobe.com/dms-mobile/bourbon-core-java-lifecycle/blob/6dbd0b666f41af458f445b94de2bc64217c1fda7/code/lifecycle/src/main/java/com/adobe/marketing/mobile/XDMLifecycleMetricsBuilder.java#L159

Comment (emdobrin) [reply AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift:103]: Ah.. my bad, checked the spec and they should be included.

File: AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift (hunk 10 above)
Comment (emdobrin) [reply AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift:52]: this needs to be updated

File: AEPLifecycle/Tests/XDMLifecycleMetricsBuilderTests.swift [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+        mockSystemInfoService.applicationName = "test-app-name"
+        mockSystemInfoService.appVersion = "test-version"
+        mockSystemInfoService.mobileCarrierName = "test-carrier"
+        mockSystemInfoService.platformName = "test-platform"
+        mockSystemInfoService.operatingSystemName = "test-os-name"
+        mockSystemInfoService.operatingSystemVersion = "test-os-version"
+        mockSystemInfoService.deviceName = "test-device-name"
+        mockSystemInfoService.deviceModelNumber = "test-device-model"
+        mockSystemInfoService.displayInformation = (100, 200)
+        mockSystemInfoService.orientation = .PORTRAIT
```
Comment (emdobrin) [inline AEPLifecycle/Tests/XDMLifecycleMetricsBuilderTests.swift:39]: nit: not used in these examples

**Review by emdobrin — APPROVED** [review]
Couple tiny suggestions, but looks good otherwise.

**Pull request merged**